File: src/digraphx/min_parmetric_q.py
from abc import abstractmethod
from fractions import Fraction
from typing import Generic, Mapping, MutableMapping, Tuple, TypeVar, Callable

from .neg_cycle_q import Cycle, Edge, NegCycleFinder, Node

Domain = TypeVar("Domain", int, Fraction, float)  # Comparable Ring
Ratio = TypeVar("Ratio", Fraction, float)


class MinParametricAPI(Generic[Node, Edge, Ratio]):
    @abstractmethod
    def distance(self, ratio: Ratio, edge: Edge) -> Ratio:
        """
        The `distance` function calculates the distance between a given ratio and edge.

        :param ratio: The `ratio` parameter is of type `Ratio`. It represents a ratio or proportion
        :type ratio: Ratio
        :param edge: The `edge` parameter represents an edge in a graph. It is of type `Edge`
        :type edge: Edge
        """
        pass

    @abstractmethod
    def zero_cancel(self, cycle: Cycle) -> Ratio:
        """
        The `zero_cancel` function takes a `Cycle` object as input and returns a `Ratio` object.

        :param cycle: The `cycle` parameter is of type `Cycle`.
        :type cycle: Cycle
        """
        pass


class MinParametricSolver(Generic[Node, Edge, Ratio]):
    """Minimum Parametric Solver

    This class solves the following parametric network problem:

        min  r
        s.t. dist[v] - dist[u] <= distrance(e, r)
             forall e(u, v) in G(V, E)

    A parametric network problem refers to a type of optimization problem that
    involves finding the optimal solution to a network flow problem as a function
    of one single parameter.
    """

    def __init__(
        self,
        digraph: Mapping[Node, Mapping[Node, Edge]],
        omega: MinParametricAPI[Node, Edge, Ratio],
    ) -> None:
        """
        The `__init__` function initializes an object with a graph and an omega parameter.

        :param digraph: digraph is a mapping of nodes to a mapping of nodes to edges. It represents a graph
            where each node is connected to other nodes through edges. The edges are represented by the
            mapping of nodes to edges

        :type digraph: Mapping[Node, Mapping[Node, Edge]]

        :param omega: The `omega` parameter is an instance of the `ParametricAPI` class. It represents
            some kind of parametric API that takes three type parameters: `Node`, `Edge`, and `Ratio`

        :type omega: ParametricAPI[Node, Edge, Ratio]
        """
        # self.ncf = NegCycleFinder(digraph)
        self.digraph = digraph
        self.omega: MinParametricAPI[Node, Edge, Ratio] = omega

    def run(
        self,
        dist: MutableMapping[Node, Domain],
        ratio: Ratio,
        update_ok: Callable[[Domain, Domain], bool],
        pick_one_only=False,
    ) -> Tuple[Ratio, Cycle]:
        """
        The `run` function takes in a distance mapping and a ratio, and iteratively finds the minimum
        ratio and corresponding cycle until the minimum ratio is greater than or equal to the input
        ratio.

        :param dist: The `dist` parameter is a mutable mapping where the keys are `Node` objects and the
            values are `Domain` objects. It represents the distance between nodes in a graph

        :type dist: MutableMapping[Node, Domain]

        :param ratio: The `ratio` parameter is a value that represents a ratio or proportion. It is used
            as a threshold or target value in the algorithm

        :type ratio: Ratio

        :param update_ok: The `update_ok` parameter is a function that determines whether an update to the
            distance `dist[vtx_v]` is allowed. It takes two arguments: the current value of `dist[vtx_v]` and
            the new value `d`. It should return `True` if the update is

        :return: The function `run` returns a tuple containing the updated ratio (`ratio`) and the cycle (`cycle`).
        """
        D = type(next(iter(dist.values())))

        def get_weight(e: Edge) -> Domain:
            return D(self.omega.distance(ratio, e))

        r_max = ratio
        c_max = []
        cycle = []
        reverse: bool = True

        ncf: NegCycleFinder[Node, Edge, Domain] = NegCycleFinder(self.digraph)

        while True:
            if reverse:
                cycles = ncf.howard_succ(dist, get_weight, update_ok)
            else:
                cycles = ncf.howard_pred(dist, get_weight, update_ok)
            for c_i in cycles:
                r_i = self.omega.zero_cancel(c_i)
                if r_max < r_i:
                    r_max = r_i
                    c_max = c_i
                    if pick_one_only:
                        break
            if r_max <= ratio:
                break

            cycle = c_max
            ratio = r_max
            reverse = not reverse
        return ratio, cycle

File: rust_ai/src/min_parmetric_q.rs
//! Min Parametric Solver with constraints
//!
//! This module defines a system for solving a specific type of network optimization
//! problem called a "minimum parametric problem." The purpose of this code is to
//! find the smallest possible value for a parameter (called a ratio) that
//! satisfies certain conditions in a graph-like structure.

use std::collections::HashMap;

use crate::neg_cycle_q::NegCycleFinderQ;
use crate::types::{Cycle, Domain, Edge, Node, RatioType};

/// Minimum Parametric API trait for distance calculation and cycle analysis.
///
/// This trait defines an interface for calculating distances and handling cycles,
/// allowing for different implementations of these operations.
pub trait MinParametricAPI<N, E, R>
where
    N: Node,
    E: Edge,
    R: RatioType,
{
    /// Calculate the parametric distance for an edge given the current ratio.
    fn distance(&self, ratio: R, edge: &E) -> R;

    /// Calculate the actual ratio for a given cycle.
    fn zero_cancel(&self, cycle: &Cycle<E>) -> R;
}

/// Minimum Parametric Solver with constraints
///
/// This struct solves a specific type of network optimization problem called a
/// "minimum parametric problem." It finds the smallest possible value for a
/// parameter (called a ratio) that satisfies certain conditions in a graph.
pub struct MinParametricQSolver<N, E, R, API>
where
    N: Node,
    E: Edge,
    R: RatioType,
    API: MinParametricAPI<N, E, R>,
{
    /// The graph structure where nodes map to neighbors and edges
    digraph: HashMap<N, HashMap<N, E>>,
    /// Parametric API for distance calculation and cycle analysis
    omega: API,
    /// Marker for unused type parameter R
    _marker: std::marker::PhantomData<R>,
}

impl<N, E, R, API> MinParametricQSolver<N, E, R, API>
where
    N: Node,
    E: Edge,
    R: RatioType,
    API: MinParametricAPI<N, E, R>,
{
    /// Initialize the solver with a graph and a parametric API.
    ///
    /// # Arguments
    ///
    /// * `digraph` - A mapping of nodes to a mapping of nodes to edges.
    /// * `omega` - A parametric API instance that provides methods for distance
    ///   calculation and cycle analysis.
    pub fn new(digraph: HashMap<N, HashMap<N, E>>, omega: API) -> Self {
        Self {
            digraph,
            omega,
            _marker: std::marker::PhantomData,
        }
    }

    /// Run the minimum parametric solver algorithm.
    ///
    /// The algorithm works by:
    /// 1. Starting with an initial ratio and distance estimates
    /// 2. Using a negative cycle finder to find cycles in the graph
    /// 3. For each cycle found, calculating a new ratio
    /// 4. Updating the minimum ratio if a smaller one is found
    /// 5. Repeating until no better ratio can be found
    ///
    /// The algorithm can switch between searching for cycles in the forward
    /// direction (successor nodes) and the backward direction (predecessor nodes).
    ///
    /// # Arguments
    ///
    /// * `dist` - Initial distance labels for nodes
    /// * `ratio` - Initial ratio value to start the search
    /// * `use_succ` - Whether to use successor relaxation (true) or predecessor relaxation (false)
    ///
    /// # Returns
    ///
    /// A tuple containing the optimal ratio and the cycle that achieves it.
    pub fn run(&self, mut dist: HashMap<N, R>, ratio: R, use_succ: bool) -> (R, Cycle<E>)
    where
        R: Domain + Clone,
    {
        // Initialize minimum ratio and cycle
        let mut r_min = ratio.clone();
        let mut c_min = Vec::new();
        let mut cycle = Vec::new();
        let mut current_ratio = ratio.clone();

        // Create a negative cycle finder instance with the graph
        let mut ncf: NegCycleFinderQ<N, E, R> = NegCycleFinderQ::new(self.digraph.clone());

        // Main algorithm loop
        loop {
            // Define a weight function that calculates distance based on current ratio
            let get_weight = |e: &E| self.omega.distance(current_ratio.clone(), e);

            // Find all negative cycles in the graph
            let cycles: Vec<Cycle<E>> = if use_succ {
                ncf.howard_succ(&mut dist, &get_weight)
            } else {
                ncf.howard_pred(&mut dist, &get_weight)
            };

            for ci in cycles {
                // Calculate the ratio that would make this cycle's total distance zero
                let ri = self.omega.zero_cancel(&ci);

                // Update minimum ratio if a smaller one is found
                if r_min > ri {
                    r_min = ri;
                    c_min = ci;
                }
            }

            // Termination condition: no better ratio found
            if r_min >= current_ratio {
                break;
            }

            // Update cycle and ratio for next iteration
            cycle = c_min.clone();
            current_ratio = r_min.clone();
        }

        (ratio, cycle)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use num_rational::Ratio;
    use std::collections::HashMap;

    struct TestAPI;

    impl MinParametricAPI<&'static str, i32, Ratio<i64>> for TestAPI {
        fn distance(&self, ratio: Ratio<i64>, _edge: &i32) -> Ratio<i64> {
            // Simple implementation for testing
            ratio.clone()
        }

                    fn zero_cancel(&self, _cycle: &Vec<i32>) -> Ratio<i64> {
                        // Simple implementation for testing
                        Ratio::new(1, 2)
                    }    }

    #[test]
    fn test_min_parametric_q_solver_new() {
        let _digraph: HashMap<&str, HashMap<&str, i32>> = HashMap::new();
        let api = TestAPI;
        let _solver: MinParametricQSolver<&str, i32, Ratio<i64>, TestAPI> = MinParametricQSolver::new(_digraph, api);
        // Just testing that it compiles and creates successfully
        assert!(true);
    }
}
